File: demos/47711/σ░ÅΦ▒åσ¡É/frontend/api/collab.js
import { request } from '../utils/api'

const PREFIX = '/collabs'
const CARD_GRADIENTS = [
	'linear-gradient(135deg,#f97316,#ef4444)',
	'linear-gradient(135deg,#667eea,#764ba2)',
	'linear-gradient(135deg,#34d399,#059669)',
	'linear-gradient(135deg,#ec4899,#be185d)'
]

function buildGradient(id) {
	return CARD_GRADIENTS[Math.abs(Number(id) || 0) % CARD_GRADIENTS.length]
}

function normalizeApplication(item = {}) {
	return {
		id: item.id,
		userId: item.user_id,
		applicantName: item.applicant_name || '',
		message: item.message || '',
		applyTime: item.apply_time ? new Date(item.apply_time).getTime() : 0
	}
}

function normalizeCollab(item = {}) {
	return {
		id: item.id,
		ownerUserId: item.owner_user_id ?? null,
		ownerName: item.owner_name || '',
		ocName: item.oc_name || '',
		emoji: item.emoji || '🌟',
		desc: item.description || '',
		status: item.status || '招募中',
		tags: Array.isArray(item.tags) ? item.tags : [],
		applicationCount: item.application_count || 0,
		applied: !!item.applied,
		ownedByMe: !!item.owned_by_me,
		createdAt: item.created_at ? new Date(item.created_at).getTime() : 0,
		applications: Array.isArray(item.applications) ? item.applications.map(normalizeApplication) : [],
		gradient: buildGradient(item.id)
	}
}

function extractList(res) {
	const list = Array.isArray(res?.data) ? res.data : []
	return list.map(normalizeCollab)
}

function extractDetail(res) {
	return normalizeCollab(res?.data || res)
}

export function collabRequest({ url = '', method = 'GET', data = null, auth = false, header = {} } = {}) {
	return request({ url: `${PREFIX}${url}`, method, data, auth, header })
}

export async function listCollabs() {
	const res = await collabRequest({ auth: true })
	return extractList(res)
}

export async function createCollab(data) {
	const res = await collabRequest({ method: 'POST', data, auth: true })
	return extractDetail(res)
}

export async function applyCollab(id, data = {}) {
	const res = await collabRequest({ url: `/${id}/apply`, method: 'POST', data, auth: true })
	return extractDetail(res)
}

export async function cancelCollabApply(id) {
	const res = await collabRequest({ url: `/${id}/apply`, method: 'DELETE', auth: true })
	return extractDetail(res)
}

File: demos/47711/σ░ÅΦ▒åσ¡É/backend/app/schemas/collab.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class CollabCreateRequest(BaseModel):
    oc_name: str = Field(min_length=1, max_length=100)
    emoji: str = Field(min_length=1, max_length=10)
    description: str = Field(min_length=5, max_length=300)
    tags: list[str] = Field(default_factory=list, max_length=8)

    @field_validator("oc_name", "emoji", "description")
    @classmethod
    def validate_text(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("字段不能为空")
        return cleaned

    @field_validator("tags")
    @classmethod
    def validate_tags(cls, value: list[str]) -> list[str]:
        tags: list[str] = []
        for tag in value:
            cleaned = tag.strip()
            if cleaned and cleaned not in tags:
                tags.append(cleaned[:20])
        return tags[:8]


class CollabApplyRequest(BaseModel):
    message: str | None = Field(default=None, max_length=200)

    @field_validator("message")
    @classmethod
    def validate_message(cls, value: str | None) -> str | None:
        if value is None:
            return None
        cleaned = value.strip()
        return cleaned or None


class CollabApplicationOut(BaseModel):
    id: int
    user_id: int
    applicant_name: str
    message: str | None = None
    apply_time: datetime

    model_config = ConfigDict(from_attributes=True)


class CollabOut(BaseModel):
    id: int
    owner_user_id: int | None = None
    owner_name: str
    oc_name: str
    emoji: str
    description: str
    status: str
    tags: list[str]
    application_count: int
    applied: bool = False
    owned_by_me: bool = False
    created_at: datetime
    applications: list[CollabApplicationOut] = Field(default_factory=list)
